File: code/JieLi_Home_Demo/NewJieliZhiNeng/App设置/关于/DebugView/CustomerCommand/CustomerCmdViewController.m
//
//  CustomerCmdViewController.m
//  NewJieliZhiNeng
//
//  Created by EzioChan on 2022/10/10.
//  Copyright © 2022 杰理科技. All rights reserved.
//

#import "CustomerCmdViewController.h"

@interface CustomerCmdViewController ()<UITextFieldDelegate>

@property(nonatomic,strong)UITextField *sendTextFixed;
@property(nonatomic,strong)UIButton *sendBtn;
@property(nonatomic,strong)UIButton *clearBtn;
@property(nonatomic,strong)UITextView *cmdTextLab;
@property(nonatomic,strong)NSString *logText;
@property(nonatomic,strong)JL_EntityM *entity;

@end

@implementation CustomerCmdViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.naviView.titleLab.text = @"自定义命令";
    _logText = @"";
    self.entity = [[JL_RunSDK sharedMe] mBleEntityM];
    [self stepUI];
    //接收自定义数据
    [JL_Tools add:kJL_MANAGER_CUSTOM_DATA Action:@selector(noteCustomData:) Own:self];
}


-(void)stepUI{
    _sendTextFixed = [UITextField new];
    _sendTextFixed.delegate = self;
    _sendTextFixed.keyboardType = UIKeyboardTypeASCIICapable;
    _sendTextFixed.returnKeyType = UIReturnKeyDone;
    _sendTextFixed.borderStyle = UITextBorderStyleRoundedRect;
    _sendTextFixed.clearButtonMode = UITextFieldViewModeWhileEditing;
    _sendTextFixed.backgroundColor = [UIColor colorFromHexString:@"#D7DADD"];
    
    [self.view addSubview:_sendTextFixed];
    [_sendTextFixed mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.naviView.mas_bottom).offset(10);
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.height.offset(60);
    }];
    
    _sendBtn = [UIButton new];
    [self.view addSubview:_sendBtn];
    _sendBtn.backgroundColor = [UIColor systemBlueColor];
    _sendBtn.layer.cornerRadius = 8;
    _sendBtn.layer.masksToBounds = true;
    [_sendBtn setTitle:@"Send" forState:UIControlStateNormal];
    [_sendBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [_sendBtn setTitleColor:[UIColor grayColor] forState:UIControlStateHighlighted];
    [_sendBtn addTarget:self action:@selector(sendBtnAction) forControlEvents:UIControlEventTouchUpInside];
    [_sendBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_sendTextFixed.mas_bottom).offset(10);
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.height.offset(40);
    }];
    
    _clearBtn = [UIButton new];
    [self.view addSubview:_clearBtn];
    _clearBtn.backgroundColor = [UIColor systemGreenColor];
    _clearBtn.layer.cornerRadius = 8;
    _clearBtn.layer.masksToBounds = true;
    [_clearBtn setTitle:@"Clear Cache Log" forState:UIControlStateNormal];
    [_clearBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [_clearBtn setTitleColor:[UIColor grayColor] forState:UIControlStateHighlighted];
    [_clearBtn addTarget:self action:@selector(clearBtnAction) forControlEvents:UIControlEventTouchUpInside];
    [_clearBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_sendBtn.mas_bottom).offset(10);
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.height.offset(40);
    }];
    
    _cmdTextLab = [UITextView new];
    _cmdTextLab.font = [UIFont systemFontOfSize:13];
    _cmdTextLab.textColor = [UIColor darkTextColor];
    _cmdTextLab.editable = NO;
    [self.view addSubview:_cmdTextLab];
    
    [_cmdTextLab mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_clearBtn.mas_bottom).offset(10);
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.bottom.equalTo(self.view.mas_safeAreaLayoutGuideBottom).offset(-20);
    }];
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(textFixedClosed)];
    [self.view addGestureRecognizer:tap];
    
}

-(void)textFixedClosed{
    [_sendTextFixed endEditing:true];
}

-(void)backBtnAction{
    [self dismissViewControllerAnimated:true completion:nil];
}

-(void)sendBtnAction{
    if(_sendTextFixed.text.length != 0){
        //发送自定义数据
        NSData *data = [JL_Tools HexToData:_sendTextFixed.text];
        [self addStrToLogText:[JL_Tools dataChangeToString:data] sr:true];
        [self.entity.mCmdManager.mCustomManager cmdCustomData:data
                                                            Result:^(JL_CMDStatus status,
                                                                     uint8_t sn, NSData * _Nullable data) {
            if (status == JL_CMDStatusSuccess) {
                kJLLog(JLLOG_DEBUG,@"发数成功...");
            }else{
                kJLLog(JLLOG_DEBUG,@"发数失败~");
            }
        }];
    }else{
        [DFUITools showText:@"Error for nothing to send" onView:self.view delay:1];
    }
}

-(void)clearBtnAction{
    _logText = @"";
    _cmdTextLab.text = _logText;
}


-(void)addStrToLogText:(NSString *)str sr:(BOOL)type{
    NSString *typeStr = @"send:";
    if(!type){
        typeStr = @"receive:";
    }
    NSDateFormatter *fm = [NSDateFormatter new];
    fm.dateFormat = @"yyyy-MM-dd hh:mm:ss";
    NSString *dateStr = [fm stringFromDate:[NSDate new]];
    NSMutableString *mstr = [NSMutableString new];
    [mstr appendFormat:@"\n%@\n%@%@\n",dateStr,typeStr,str];
    [mstr appendString:_logText];
    _logText = mstr;
    _cmdTextLab.text = _logText;
    [_cmdTextLab scrollsToTop];
}


//MARK: - received data from device
-(void)noteCustomData:(NSNotification *)note{
    NSData *data = note.object;
    NSString *recStr = [JL_Tools dataChangeToString:data];
    [self addStrToLogText:recStr sr:false];
}

-(BOOL)textFieldShouldReturn:(UITextField *)textField{
    [textField endEditing:true];
    return true;
}

@end

File: code/JieLi_Home_Demo/NewJieliZhiNeng/GlobalViews/AppSettingVC/AboutVC/DebugView/CustomerCommand/CustomerCmdViewController.swift
import UIKit
import SnapKit
import JL_BLEKit
import Combine

class CustomerCmdViewController: DebugBasicViewController, UITextFieldDelegate {
    
    private var sendTextFixed: UITextField!
    private var sendBtn: UIButton!
    private var clearBtn: UIButton!
    private var cmdTextLab: UITextView!
    private var logText: String = ""
    private var entity: JL_EntityM?
    private var cancellables = Set<AnyCancellable>()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationView.title = "自定义命令"
        logText = ""
        entity = JL_RunSDK.sharedMe().mBleEntityM
        stepUI()
        
        // 使用 Combine 接收自定义数据
        NotificationCenter.default.publisher(for: NSNotification.Name(kJL_MANAGER_CUSTOM_DATA))
            .sink { [weak self] note in
                guard let data = note.object as? Data else { return }
                let recStr = JL_Tools.dataChange(toString: data) ?? ""
                self?.addStrToLogText(recStr, sr: false)
            }
            .store(in: &cancellables)
    }
    
    private func stepUI() {
        subTable?.isHidden = true
        
        sendTextFixed = UITextField()
        sendTextFixed.delegate = self
        sendTextFixed.keyboardType = .asciiCapable
        sendTextFixed.returnKeyType = .done
        sendTextFixed.borderStyle = .roundedRect
        sendTextFixed.clearButtonMode = .whileEditing
        sendTextFixed.backgroundColor = UIColor(fromHexString: "#D7DADD")
        
        view.addSubview(sendTextFixed)
        sendTextFixed.snp.makeConstraints { make in
            make.top.equalTo(navigationView.snp.bottom).offset(10)
            make.left.right.equalToSuperview().inset(20)
            make.height.equalTo(60)
        }
        
        sendBtn = UIButton()
        view.addSubview(sendBtn)
        sendBtn.backgroundColor = .systemBlue
        sendBtn.layer.cornerRadius = 8
        sendBtn.layer.masksToBounds = true
        sendBtn.setTitle("Send", for: .normal)
        sendBtn.setTitleColor(.white, for: .normal)
        sendBtn.setTitleColor(.gray, for: .highlighted)
        sendBtn.addTarget(self, action: #selector(sendBtnAction), for: .touchUpInside)
        
        sendBtn.snp.makeConstraints { make in
            make.top.equalTo(sendTextFixed.snp.bottom).offset(10)
            make.left.right.equalToSuperview().inset(20)
            make.height.equalTo(40)
        }
        
        clearBtn = UIButton()
        view.addSubview(clearBtn)
        clearBtn.backgroundColor = .systemGreen
        clearBtn.layer.cornerRadius = 8
        clearBtn.layer.masksToBounds = true
        clearBtn.setTitle("Clear Cache Log", for: .normal)
        clearBtn.setTitleColor(.white, for: .normal)
        clearBtn.setTitleColor(.gray, for: .highlighted)
        clearBtn.addTarget(self, action: #selector(clearBtnAction), for: .touchUpInside)
        
        clearBtn.snp.makeConstraints { make in
            make.top.equalTo(sendBtn.snp.bottom).offset(10)
            make.left.right.equalToSuperview().inset(20)
            make.height.equalTo(40)
        }
        
        cmdTextLab = UITextView()
        cmdTextLab.font = UIFont.systemFont(ofSize: 13)
        cmdTextLab.textColor = .darkText
        cmdTextLab.isEditable = false
        view.addSubview(cmdTextLab)
        
        cmdTextLab.snp.makeConstraints { make in
            make.top.equalTo(clearBtn.snp.bottom).offset(10)
            make.left.right.equalToSuperview().inset(20)
            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).offset(-20)
        }
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(textFixedClosed))
        view.addGestureRecognizer(tap)
    }
    
    @objc private func textFixedClosed() {
        sendTextFixed.endEditing(true)
    }
    
    @objc private func sendBtnAction() {
        if let text = sendTextFixed.text, !text.isEmpty {
            let data = JL_Tools.hex(toData: text) as? Data ?? Data()
            addStrToLogText(JL_Tools.dataChange(toString: data) ?? "", sr: true)
            
            entity?.mCmdManager.mCustomManager.cmdCustomData(data, result: { status, sn, resData in
                if status == .success {
                    JLLogManager.logLevel(.DEBUG, content: "发数成功...")
                } else {
                    JLLogManager.logLevel(.DEBUG, content: "发数失败~")
                }
            })
        } else {
            DFUITools.showText("Error for nothing to send", on: self.view, delay: 1.0)
        }
    }
    
    @objc private func clearBtnAction() {
        logText = ""
        cmdTextLab.text = logText
    }
    
    private func addStrToLogText(_ str: String, sr: Bool) {
        let typeStr = sr ? "send:" : "receive:"
        let fm = DateFormatter()
        fm.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let dateStr = fm.string(from: Date())
        
        var mstr = ""
        mstr.append("\n\(dateStr)\n\(typeStr)\(str)\n")
        mstr.append(logText)
        logText = mstr
        cmdTextLab.text = logText
        
        let range = NSRange(location: 0, length: 1)
        cmdTextLab.scrollRangeToVisible(range)
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.endEditing(true)
        return true
    }
}
